numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -4189,10 +4189,80 @@ static PyNumberMethods clongdoubletype_as_number;
+    0,                                          /* tp_setattr */
+#if defined(NPY_PY3K)
+    (void *)0,                                  /* tp_reserved */
+#else
+    0,                                          /* tp_compare */
+#endif
+    (reprfunc)generic_meta_repr,                /* tp_repr */
+    0,                                          /* tp_as_number */
+    0,                                          /* tp_as_sequence */
+    0,                                          /* tp_as_mapping */
+    0,                                          /* tp_hash */
+    0,                                          /* tp_call */
+    0,                                          /* tp_str */
+    0,                                          /* tp_getattro */
+    0,                                          /* tp_setattro */
+    0,                                          /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT,                         /* tp_flags */
+    0,                                          /* tp_doc */
+    0,                                          /* tp_traverse */
+    0,                                          /* tp_clear */
+    0,                                          /* tp_richcompare */
+    0,                                          /* tp_weaklistoffset */
+    0,                                          /* tp_iter */
+    0,                                          /* tp_iternext */
+    0,                                          /* tp_methods */
+    0,                                          /* tp_members */
+    0,                                          /* tp_getset */
+    0,                                          /* tp_base */
+    0,                                          /* tp_dict */
+    0,                                          /* tp_descr_get */
+    0,                                          /* tp_descr_set */
+    0,                                          /* tp_dictoffset */
+    0,                                          /* tp_init */
+    0,                                          /* tp_alloc */
+    0,                                          /* tp_new */
+    0,                                          /* tp_free */
+    0,                                          /* tp_is_gc */
+    0,                                          /* tp_bases */
+    0,                                          /* tp_mro */
+    0,                                          /* tp_cache */
+    0,                                          /* tp_subclasses */
+    0,                                          /* tp_weaklist */
+    0,                                          /* tp_del */
+    0,                                          /* tp_version_tag */
+};
+
+
 NPY_NO_EXPORT void
 initialize_numeric_types(void)
 {
     init_basetypes();
+
+    PyGeneric_MetaType.tp_base = &PyType_Type;
+    PyType_Ready(&PyGeneric_MetaType);
+
+    ((PyObject *)&PyGenericArrType_Type)->ob_type = &PyGeneric_MetaType;
```

Comment:
CPython warns against setting this statically due to some compiler bug, I think.

Reply:
Oh, interesting, that is good to know. Using `PyVarObject_HEAD_INIT(&PyGeneric_MetaType, 0)` is not safe? I was not aware of that, have to try and remember that :).
The main purpose of this to make the multiple inheritance saner? That said, I can see a MetaType make a lot of sense. Although, I can also see some sense in going the duck route for scalars (well, for scalars created by users).

Reply:
> Using PyVarObject_HEAD_INIT(&PyGeneric_MetaType, 0) is not safe? 

From what I remember, MSVC had a bug where `&PyGeneric_MetaType` doesn't resolve correctly in a static initializer. I unfortunately don't have the reference to hand.

> The main purpose of this to make the multiple inheritance saner

The original motivation here was just to override `type(np.uint8).__repr__` so that it could show something like `<type 'np.uint8' (C char)>`, to solve aliasing confusion.